keylime/dsse/x509.py

```diff
@@ -0,0 +1,109 @@
+"""x509 signing/verification implementation.
+SPDX-License-Identifier: Apache-2.0
+"""
+
+import base64
+import datetime
+
+from cryptography import x509
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+from cryptography.x509 import Certificate
+from cryptography.x509.oid import NameOID
+
+from . import dsse
+
+
+class Signer(dsse.Signer):
+    def __init__(self, secret_key: ec.EllipticCurvePrivateKey):
+        self.secret_key = secret_key
+        self.certificate = self.construct(secret_key)
+
+    @classmethod
+    def create(cls, keypath: str = "private_x509.key") -> "Signer":
+        private_key = ec.generate_private_key(ec.SECP256R1(), default_backend())
+        with open(keypath, "wb") as pem_out:
+            pem_out.write(
+                private_key.private_bytes(
+                    encoding=serialization.Encoding.PEM,
+                    format=serialization.PrivateFormat.TraditionalOpenSSL,
+                    encryption_algorithm=serialization.NoEncryption(),
+                )
+            )
+        return Signer(private_key)
+
+    @classmethod
+    def construct(
+        cls,
+        private_key: ec.EllipticCurvePrivateKey,
+        subject_name: str = "dsse_lib",
+        issuer_name: str = "dsse_lib",
+        subject_alternative_name: str = "dsse_lib",
+        expiration: int = 30,
+    ) -> Certificate:
+        one_day = datetime.timedelta(1, 0, 0)
+        public_key = private_key.public_key()
+        builder = x509.CertificateBuilder()
+        builder = builder.subject_name(
+            x509.Name(
+                [
+                    x509.NameAttribute(NameOID.COMMON_NAME, subject_name),
+                ]
+            )
+        )
+        builder = builder.issuer_name(
+            x509.Name(
+                [
+                    x509.NameAttribute(NameOID.COMMON_NAME, issuer_name),
+                ]
+            )
+        )
+        builder = builder.not_valid_before(datetime.datetime.today() - one_day)
+        builder = builder.not_valid_after(datetime.datetime.today() + (one_day * expiration))
+        builder = builder.serial_number(x509.random_serial_number())
+        builder = builder.public_key(public_key)
+        builder = builder.add_extension(
+            x509.SubjectAlternativeName([x509.DNSName(subject_alternative_name)]), critical=False
+        )
+        builder = builder.add_extension(
+            x509.BasicConstraints(ca=False, path_length=None),
+            critical=True,
+        )
+        certificate = builder.sign(
+            private_key=private_key,
+            algorithm=hashes.SHA256(),
+        )
+
+        with open("certificate.crt", "wb") as crt_out:
```

Comment:
It writes this file and I think the help description for the x509 backend should describe it.

Reply:
@stefanberger Where would you recommend I put documentation for that?

Reply:
I think you should mention it as part of the x509 backend though the help for the "--backend' option seems also not such a great place. Another possibility would be to introduce an option that gives the user control over the name of the certificate file, just like the name of the key file.